core/broadcaster.go

```diff
@@ -0,0 +1,117 @@
+package core
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"net/http"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/golang/glog"
+
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/livepeer/go-livepeer/common"
+	ethTypes "github.com/livepeer/go-livepeer/eth/types"
+	ffmpeg "github.com/livepeer/lpms/ffmpeg"
+)
+
+var ErrNotFound = errors.New("ErrNotFound")
+
+// Broadcaster RPC interface implementation
+
+type broadcaster struct {
+	node  *LivepeerNode
+	httpc *http.Client
+	job   *ethTypes.Job
+	tinfo interface{}
+}
+
+func (bcast *broadcaster) Sign(msg []byte) ([]byte, error) {
+	if bcast.node == nil || bcast.node.Eth == nil {
+		return []byte{}, fmt.Errorf("Cannot sign; missing eth client")
+	}
+	return bcast.node.Eth.Sign(crypto.Keccak256(msg))
+}
+func (bcast *broadcaster) Job() *ethTypes.Job {
+	return bcast.job
+}
+func (bcast *broadcaster) GetHTTPClient() *http.Client {
+	return bcast.httpc
+}
+func (bcast *broadcaster) SetHTTPClient(hc *http.Client) {
+	bcast.httpc = hc
+}
+func (bcast *broadcaster) GetTranscoderInfo() interface{} {
+	return bcast.tinfo
+}
+func (bcast *broadcaster) SetTranscoderInfo(t interface{}) {
+	bcast.tinfo = t
+}
+func NewBroadcaster(node *LivepeerNode, job *ethTypes.Job) *broadcaster {
+	return &broadcaster{
+		node: node,
+		job:  job,
+	}
+}
+
+//CreateTranscodeJob creates the on-chain transcode job.
+func (n *LivepeerNode) CreateTranscodeJob(strmID StreamID, profiles []ffmpeg.VideoProfile, price *big.Int) (*ethTypes.Job, error) {
```

Comment:
Having `LivepeerNode` methods in `broadcaster.go` feels a little misleading.  I think we should either keep the `CreateTranscodeJob` method in the `livepeernode.go` file, or create a `livepeernode` package that includes multiple files that only contain `LivepeerNode`-specific methods (for example, we can have `transcodeJob.go` in `/core/livepeernode/`).  It feels like the "job creation" responsibility should belong to `LivepeerNode`.

Reply:
Part of the idea was to keep functionally related code together, and unrelated code separate. Given the current setup in `core` :

* transcoder.go : Code that is meant to be called only by the transcoder / orchestrator. That's `TranscodeSegment` related and the RPC interface implementation.
* broadcaster.go : Code that is meant to be called only by the broadcaster. `CreateTranscodeJob` is the major one here (transcoders don't create jobs), in addition to the RPC interface impl.
* `livepeernode.go` - Shared code common to both node types. There is very little. Looking at what we have left in livepeernode, `GetClaimManager` probably should also be moved to the transcoder/orchestrator. Many of the LivepeerNode fields are also transcoder-specific.

This separation could be taken further. We can remove `LivepeerNode` as a standalone entity, and create a  `BroadcasterNode` and `TranscoderNode` .  The livepeernode struct can be kept, and the broadcaster/transcoder can inherit common fields. We might have to tweak StartEthServices/StopEthServices (the only shared functions right now) but that is relatively minor.

Personally, I feel that the implicit context given by the filename provides enough of a clue as to where/how the code should be used, regardless of the underlying 'base struct'. We can go too far in subdividing code, eg files implementing only one function such as `transcodeJob.go`. But that boils down to preferences.

Reply:
I see.  That seems logical - can we add this logical layout as a comment in `livepeernode.go` for documentation?

Reply:
Yep, comments added here https://github.com/livepeer/go-livepeer/pull/512/commits/84e0261776b74d862879ec4fd8a54ee336228896